operator/pkg/legacy/secrets.go

```diff
@@ -0,0 +1,157 @@
+// a helm values table keyed with "key".
+//
+// However, unlike with the common key and secrets supplied to the constructor, this one will be given the following
+// special treatment. Before injection, the list of secrets provided in this call is first extended by prepending
+// the list of secrets in vals[common key].useExisting (i.e. in the *input* helm values), if any.
+//
+// This is done such that the useExisting secrets in the table at the common key get properly propagated to the extra table.
+func (i *injector) WithExtraImagePullSecrets(key string, secrets ...string) *injector {
+	if i.extraSecretMap == nil {
+		i.extraSecretMap = map[string][]string{key: secrets}
+	} else {
+		i.extraSecretMap[key] = secrets
+	}
+	return i
+}
+
+type injector struct {
+	client          ctrlClient.Reader
+	commonTableName string
+	commonSecrets   []string
+	extraSecretMap  map[string][]string
+}
+
+var _ translation.Enricher = &injector{}
+
+// Enrich modifies vals to append - for each {key,secrets} entry in secret map - secrets to vals[key]["useExisting"].
+func (i *injector) Enrich(ctx context.Context, obj k8sutil.Object, vals chartutil.Values) (chartutil.Values, error) {
+	namespaceName := obj.GetNamespace()
+	upstreamCommonSecrets, err := getUseExisting(vals, i.commonTableName)
+	if err != nil {
+		return nil, err
+	}
+	vals, err = i.enrich(ctx, vals, i.commonTableName, i.commonSecrets, namespaceName)
+	if err != nil {
+		return nil, err
+	}
+	for key, secretNames := range i.extraSecretMap {
+		vals, err = i.enrich(ctx, vals, key, append(upstreamCommonSecrets, secretNames...), namespaceName)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return vals, nil
+}
+
+// getUseExisting returns a slice of secret names at vals.key.useExisting.
+// If no secret names are specified, returns an empty slice and nil error.
+// However, if an unexpected type is encountered, returns an error.
+func getUseExisting(vals chartutil.Values, key string) ([]string, error) {
+	parentValue, ok := vals[key]
+	if !ok {
+		return nil, nil
+	}
+	parentTable, ok := parentValue.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("key %q maps to a %T, table expected", key, parentValue)
+	}
+	value, ok := parentTable["useExisting"]
+	if !ok {
+		return nil, nil
+	}
+	secretsSlice, ok := value.([]interface{})
+	if !ok {
+		return nil, fmt.Errorf("unexpected value %q type: %T", key+".useExisting", value)
+	}
+	var secretStringSlice []string
+	for i, e := range secretsSlice {
+		var secret string
+		if secret, ok = e.(string); ok {
+			secretStringSlice = append(secretStringSlice, secret)
+		} else {
+			return nil, fmt.Errorf("unexpected %q element %d type: %T", key+".useExisting", i, e)
+		}
+	}
+	return secretStringSlice, nil
+}
+
+func (i *injector) enrich(ctx context.Context, vals chartutil.Values, key string, secretNames []string, namespaceName string) (chartutil.Values, error) {
+	var secretNamesToAdd []string
```

Comment:
suggestion:
```
secretNamesToAdd := make([]string, 0, len(secretNames))
```

Reply:
Actually here I'd optimize for the common case of all these secrets being absent, i.e. empty slice at the end.